File: BinaryTree/ts/maxPathSum.ts
import { TreeNode } from "./TreeNode";

export function maxPathSum(root: TreeNode<number> | null): number {
  if (root === null) {
    return Number.NEGATIVE_INFINITY;
  }
  if (root.left === null && root.right === null) {
    return root.val;
  }
  return root.val + Math.max(maxPathSum(root.left), maxPathSum(root.right));
}

File: BinaryTree/go/maxPathSum.go
package binarytree

// MaxPathSum returns the maximum root-to-leaf path sum.
func MaxPathSum(root *TreeNode) int {
	if root == nil {
		return minInt
	}
	if root.Left == nil && root.Right == nil {
		return root.Val
	}

	left := MaxPathSum(root.Left)
	right := MaxPathSum(root.Right)
	if left > right {
		return root.Val + left
	}
	return root.Val + right
}
